### ISSUE TITLE
Inconsistency Between Quantization Loss Implementation and Paper Description

### ISSUE THREAD

**Issue by cheguevara-great-man** [Marsrocky/EfficientFi#4]
Hello,

I have noticed some discrepancies between the implementation of the quantization loss in the code and the mechanism described in the paper, which may lead to incorrect gradient flow. The issues are as follows:

1. **Lack of a Stop-Gradient Operation:**  
   The paper describes an approach similar to:
   ```python
   quantized = inputs + (quantized - inputs).detach()
   ```
   to ensure that during backpropagation the gradient from the quantization branch (i.e., the `(quantized - inputs)` part) is blocked, so that only the identity connection passes the gradient to the encoder. However, in the code, it is written as:
   ```python
   quantized = inputs + (quantized - inputs)
   ```
   This allows gradients to flow both through the quantization branch and the input, leading to updates in both the encoder and the codebook. This behavior does not match the intended "straight-through" mechanism described in the paper.

2. **Loss Function Calculation Issue:**  
   In the Quantize module, the code uses:
   ```python
   e_latent_loss = F.mse_loss(quantized, inputs)
   q_latent_loss = F.mse_loss(quantized, inputs)
   loss = q_latent_loss + self._commitment_cost * e_latent_loss
   ```
   Here, both `e_latent_loss` and `q_latent_loss` are computed using the same inputs, and without any gradient blocking, causing gradients to flow from the quantization branch back to the encoder. This contradicts the desired behavior where the gradient should not pass from \( E_d(x) \) to \( E_c(x) \).

**Expected Behavior:**  
According to the VQ-VAE principles and the paper description, I suggest modifying the implementation to something like:

```python
e_latent_loss = F.mse_loss(quantized.detach(), inputs)
q_latent_loss = F.mse_loss(quantized, inputs.detach())
loss = q_latent_loss + self._commitment_cost * e_latent_loss
quantized = inputs + (quantized - inputs).detach()
```

This ensures that:
- The forward pass output remains `quantized`.
- During backpropagation, the `(quantized - inputs)` part is detached so that the gradient flows only through the identity path to the encoder, preventing incorrect updates to both the codebook and the encoder.

Could you please confirm whether this is an oversight or an intentional design choice? I believe correcting this could help avoid potential training instability.

Thank you for your excellent work, and I look forward to your response!

— Xiaoyang

---
